jwt/transport/transport.go

```diff
@@ -0,0 +1,175 @@
+		rt = http.DefaultTransport
+	}
+
+	t := &ImplWithJWT{
+		transport: rt,
+	}
+
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+
+		if err := opt(t); err != nil {
+			return nil, err
+		}
+	}
+
+	// if expire is not set or negative value, default to 5 days.
+	if t.expire < 1 {
+		t.expire = time.Duration(time.Hour * 24 * 5)
+	}
+
+	if err := t.GenerateToken(); err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// GenerateToken is a function that generates JWT and makes a POST request to bouncer to sign it.
+func (t *ImplWithJWT) GenerateToken() error {
+	t.Lock()
+	defer t.Unlock()
+
+	// TODO: this is very broken with the latest `jwt-go` lib version 3.0.0
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Claims["uid"] = t.uid
+	token.Claims["exp"] = time.Now().Add(t.expire).Unix()
+
+	tokenStr, err := token.SignedString([]byte(t.secret))
+	if err != nil {
+		return err
+	}
+
+	if tokenStr == "" {
+		return errors.New("Empty token")
+	}
+
+	authReq := struct {
+		UID   string `json:"uid"`
+		Token string `json:"token,omitempty"`
+		Exp   int64  `json:"exp,omitempty"`
+	}{
+		UID:   t.uid,
+		Token: tokenStr,
+		Exp:   time.Now().Add(t.expire).Unix(),
+	}
+
+	b, err := json.Marshal(authReq)
+	if err != nil {
+		return err
+	}
+
+	authBody := bytes.NewBuffer(b)
+	req, err := http.NewRequest("POST", t.loginEndpoint, authBody)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-type", "application/json")
+	resp, err := t.transport.RoundTrip(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return ErrTokenRefresh{
+			msg: fmt.Sprintf("POST %s failed, expect response code 200. Got %d", t.loginEndpoint, resp.StatusCode),
+		}
+	}
+
+	var authResp struct {
+		Token string `json:"token"`
+	}
+
+	if err = json.NewDecoder(resp.Body).Decode(&authResp); err != nil {
+		return err
+	}
+
+	t.token = authResp.Token
+	return nil
+}
+
+// CurrentToken returns a current JWT.
+func (t *ImplWithJWT) CurrentToken() string {
+	t.Lock()
+	defer t.Unlock()
+	return t.token
+}
+
+// RoundTrip is implementation of RoundTripper interface.
+func (t *ImplWithJWT) RoundTrip(req *http.Request) (*http.Response, error) {
+	// helper function to update `Authorization` header.
+	addAuthToken := func() {
+		if token := t.CurrentToken(); token != "" {
+			req.Header.Set("Authorization", "token="+token)
+		}
+	}
+
+	var (
+		resp *http.Response
+		err  error
+	)
+
+	// try first time
+	addAuthToken()
+	resp, err = t.transport.RoundTrip(req)
+	if err != nil {
+		return resp, err
+	}
+
+	// if request returned 401 retry one more time.
+	if resp.StatusCode == http.StatusUnauthorized {
+		if err := t.GenerateToken(); err != nil {
+			return resp, err
+		}
+
+		addAuthToken()
+		resp, err = t.transport.RoundTrip(req)
+	}
+	return resp, nil
+}
+
+// SetTransport allows to update the currently used real transport.
+func (t *ImplWithJWT) SetTransport(rt http.RoundTripper) {
```

Comment:
why is this needed at all? feels dangerous to allow internal `transport` to be swapped out at runtime. if it's really needed, should be trivial to add later. I vote to leave this OUT for now.